graphql/batch_executor.go

```diff
@@ -0,0 +1,409 @@
+// to control how we traverse the Execution graph.  Examples would include using
+// a bounded goroutine pool, or using unbounded goroutine generation for each
+// execution unit.
+type ExecutionScheduler interface {
+	Run(resolver UnitResolver, startingUnits ...*ExecutionUnit)
+}
+
+func NewBatchExecutor(scheduler ExecutionScheduler) *BatchExecutor {
+	return &BatchExecutor{
+		scheduler: scheduler,
+	}
+}
+
+// BatchExecutor is a GraphQL executor.  Given a query it can run through the
+// execution of the request.
+type BatchExecutor struct {
+	scheduler ExecutionScheduler
+}
+
+// Execute executes a query by traversing the GraphQL query graph and resolving
+// or executing fields.  Any work that needs to be done is passed off to the
+// scheduler to handle managing concurrency of the request.
+// It must return a JSON marshallable response (or an error).
+func (e *BatchExecutor) Execute(ctx context.Context, typ Type, source interface{}, query *Query) (interface{}, error) {
+	queryObject, ok := typ.(*Object)
+	if !ok {
+		return nil, fmt.Errorf("expected query or mutation object for execution, got: %s", typ.String())
+	}
+
+	topLevelSelections := Flatten(query.SelectionSet)
+	topLevelRespWriter := newTopLevelObjectWriter(query.Name)
+	initialSelectionExecUnits := make([]*ExecutionUnit, 0, len(topLevelSelections))
+	writers := make(map[string]*objectWriter)
+	for _, selection := range topLevelSelections {
+		field, ok := queryObject.Fields[selection.Name]
+		if !ok {
+			return nil, fmt.Errorf("invalid top-level selection %q", selection.Name)
+		}
+
+		writer := newObjectWriter(topLevelRespWriter, selection.Alias)
+		writers[selection.Alias] = writer
+
+		initialSelectionExecUnits = append(
+			initialSelectionExecUnits,
+			&ExecutionUnit{
+				ctx:          ctx,
+				sources:      []interface{}{source},
+				field:        field,
+				destinations: []*objectWriter{writer},
+				selection:    selection,
+			},
+		)
+	}
+
+	e.scheduler.Run(resolveExecutionUnit, initialSelectionExecUnits...)
+
+	if topLevelRespWriter.errRecorder.err != nil {
+		return nil, topLevelRespWriter.errRecorder.err
+	}
+	return writers, nil
+}
+
+// resolveExecutionUnit executes/resolves an execution unit and checks the
+// selections of the unit to determine if it needs to schedule more work (which
+// will be returned as new execution units that will need to get scheduled.
+func resolveExecutionUnit(unit *ExecutionUnit) []*ExecutionUnit {
+	// TODO Insert Batching code execution here if field is "Batch"
+
+	var units []*ExecutionUnit
+	for idx, src := range unit.sources {
+		units = append(units, safeResolveNonBatchUnit(src, unit.destinations[idx], unit)...)
+	}
+	return units
+}
+
+// safeResolveNonBatchUnit resolves a non-batch unit within a panic wrapper.
+func safeResolveNonBatchUnit(src interface{}, dest *objectWriter, unit *ExecutionUnit) []*ExecutionUnit {
+	defer func() {
+		if panicErr := recover(); panicErr != nil {
```

Comment:
share this with other code?

Reply:
Moved the wrapper to be around the field.Resolve func directly.  (Nothing else should panic).